2020/5/boarding_passes/src/main.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::HashSet;
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead};
+
+fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+    let file = File::open(&args[1])?;
+    let reader = io::BufReader::new(file);
+
+    // if let Ok(max) = max_id(reader) {
+    //     println!("The largest id is {}", max);
+    // }
+
+    let mut all_passes = HashSet::new();
+    for line in reader.lines() {
+        if let Ok(l) = line {
+            let bp = BoardingPass::new(&l)?;
+            all_passes.insert(bp.seat_id());
+        }
+    }
+
+    for row in 1..127 {
+        for col in 0..7 {
+            let id = 8 * row + col;
+            if !all_passes.contains(&id)
+                && all_passes.contains(&(id - 1))
+                && all_passes.contains(&(id + 1))
+            {
+                println!("Your seat id is {}", id);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn max_id(reader: io::BufReader<std::fs::File>) -> Result<usize, Error> {
+    let mut max = 0;
+    for line in reader.lines() {
+        if let Ok(l) = line {
+            let bp = BoardingPass::new(&l)?;
+            let id = bp.seat_id();
+            if id > max {
+                max = id
+            }
+        }
+    }
+
+    Ok(max)
+}
+
+pub struct BoardingPass {
+    definition: String,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    InvalidBoardingPass,
+    IOError(std::io::Error),
+}
+
+impl std::convert::From<std::io::Error> for Error {
```

Comment:
I quite like this error handling/wrapping pattern. Though I wish it were easier to automatically capture a "stack trace" (either literal or some other "conceptual" trace of the path that led to the error). Once you have a bunch of functions calling each other and all using the same Error type it gets messy.

There are some crates out there that try to do something but I haven't yet found a great pattern. I like returning errors as an alternative to exceptions (and Rust has way better syntax for it than Go) but I still miss stack traces :(